File: Backpack_II.cpp
class Solution {
public:
    /**
     * @param m: An integer m denotes the size of a backpack
     * @param A & V: Given n items with size A[i] and value V[i]
     * @return: The maximum value
     */
    int backPackII(int m, vector<int> A, vector<int> V) {
        // write your code here
        //table[i][j] denotes the maxVal the backpack can get from the first i elements
        vector<vector<int>> table(2, vector<int>(m + 1, 0));
        for (int i = 1; i < A.size() + 1; ++i){
            for (int j = 1; j < m + 1; ++j){
                table[i % 2][j] = table[(i - 1) % 2][j];
                if (j >= A[i - 1]) table[i % 2][j] = max(table[(i - 1) % 2][j - A[i - 1]] + V[i - 1], table[i % 2][j]);
            }
        }
        return table[A.size() % 2][m];
    }
};

File: Backpack II.py
class Solution:
    # @param m: An integer m denotes the size of a backpack
    # @param A & V: Given n items with size A[i] and value V[i]
    # @return: The maximum value
    def backPackII(self, m, A, V):
        # write your code here
        
        length = len(A)
        dp = [[0] * (m + 1) for __ in range(2)]
        
        for i in range(1, length + 1):
            for j in range(1, m + 1):
                dp[i % 2][j] = dp[(i - 1) % 2][j]
                if A[i - 1] <= j:
                    dp[i % 2][j] = max(dp[i % 2][j], dp[(i - 1) % 2][j - A[i - 1]] + V[i - 1])
        return dp[length % 2][m]